ReactQt/tests/test-array-reconciliation.cpp

```diff
@@ -0,0 +1,163 @@
+/**
+ * Copyright (c) 2017-present, Status Research and Development GmbH.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <QSignalSpy>
+#include <QTest>
+#include <QTimer>
+#include <QtQuick/QQuickView>
+
+#include "reacttestcase.h"
+#include "redbox.h"
+
+class TestArrayReconciliation : public ReactTestCase {
+    Q_OBJECT
+
+private slots:
+    virtual void initTestCase() override;
+    CLEANUP_TEST_CASE_DEFAULT(ReactTestCase)
+
+    void testComponentsArrayFirstElementInsert();
+    void testComponentsArrayLastElementDelete();
+};
+
+void TestArrayReconciliation::initTestCase() {
+    ReactTestCase::initTestCase();
+    loadQML(QUrl("qrc:/TestArrayReconciliation.qml"));
+    waitAndVerifyJsAppStarted();
+    showView();
+}
+
+void TestArrayReconciliation::testComponentsArrayFirstElementInsert() {
+    const int INITIAL_ITEMS_COUNT = 3;
+    const int CLICKED_ITEMS_COUNT = 2;
+    QQuickItem* topView = topJSComponent();
+    QCOMPARE(valueOfControlProperty(topView, "objectName").toString(), QString("topView"));
+
+    QCOMPARE(topView->childItems().size(), INITIAL_ITEMS_COUNT);
+
+    QQuickItem* textInput = topView->childItems().at(0);
+    QCOMPARE(valueOfControlProperty(textInput, "p_title").toString(), QString("FirstButton"));
+    QCOMPARE(valueOfControlProperty(topView->childItems().at(1), "p_title").toString(), QString("SecondButton"));
+    QCOMPARE(valueOfControlProperty(topView->childItems().at(2), "p_title").toString(), QString("ThirdButton"));
+
+    clickItem(textInput);
+
+    QTimer* clickDelayTimer = new QTimer(this);
+    clickDelayTimer->setSingleShot(true);
+    clickDelayTimer->setInterval(1000);
+    clickDelayTimer->start();
+    waitAndVerifyCondition([=]() { return !clickDelayTimer->isActive(); }, "Timer timeout was not triggered");
+    waitAndVerifyCondition([=]() { return topView->childItems().size() == CLICKED_ITEMS_COUNT; },
+                           "Wrong array items count after 1st click");
+
+    topView = topJSComponent();
+    textInput = topView->childItems().at(0);
+    clickItem(textInput);
+
+    clickDelayTimer->start();
```

Comment:
These 3 lines of code (starting timer and 2 waitAndVerify calls) can be extracted into a separate function to reduce code duplication in the file.

Reply:
Yep, sure, thanks